### PR TITLE
Test harness jobs

### LINKED ISSUE #6841: TestHarness should respect slots for jobs that use multiple slots
### Description of the enhancement or error report

Right now the TestHarness treats every job equally. When running parallel or threaded tests though, you can easily overload your box if you don't lower your slot request appropriately. I'd like to just make the TestHarness handle this for the user so when you say you want n slots, it'll really respect that number.

### Rationale for the enhancement or information for reproducing the error

New Test boxes are timing out a lot. This may or may not help but should be a good enhancement anyway.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

New enhancement, but impacts everyone.

### PR DESCRIPTION
**Pull request by permcody**
Add new feature to RunParallel that makes the TestHarness respect slots.

If you run without _any_ job request. All jobs will still be run regardless of size but serially.

closes #6841

### CODE REVIEW

File: python/TestHarness/testers/Tester.py [hunk 1]
```diff
@@ -80,6 +80,11 @@ def checkRunnable(self, options):
   def prepare(self):
     return
 
+  def getThreads(self, options):
+    return 1
+
+  def getProcs(self, options):
+    return 1
```
Comment (permcody) [inline python/TestHarness/testers/Tester.py]: Turns out that this isn't right for the JacobianAnalyzer. That should probably be fixed.

**Comment by permcody** [comment]
I highly encourage people to pound on this code. The RunParallel logic is ultra complicated but I've run several test scenarios on my box and this seems to be working as expected.

**Comment by YaqiWang** [comment]
So if a test requires 8 processors each with 2 threads and it sees that the number of processors idling is less than 16, this test will not be scheduled, right? Also will this test be unlikely scheduled if other tests are requiring only 1 processor and we run the test with `-j16`?

**Comment by permcody** [comment]
@YaqiWang, you've mostly got it. What the TestHarness does (always has) is queue jobs that should be runnable, but can't run immediately due to prereqs. This PR takes that further and also queues jobs that are too big to fit "currently". So in your scenario if you have a bunch of serial jobs and one 16 slot job in the middle, it'll queue that job and move on. When the jobs slots are all full the TestHarness goes into a "clean up" mode and reports the status of all finished jobs at once before queueing new ones. If in a single pass all jobs are freed up in your scenario, then your 16 slot job will run. There's always the chance that it may not run until the very end when the queue is all dried up.

There is one more feature to note here. If you don't run with any "-j" at all the queue system will run all jobs one at a time regardless of size. This is so people that are just getting their feet wet don't have skipped jobs due to insufficient slots. If you supply -j1, then any job requiring two slots will not run at all.

I've tried to capture all of the features we need.

**Pull request merged**